### ISSUE TITLE
[2pt] RFC: Topology Management

### ISSUE THREAD

**Issue by dokshina** [tarantool/cartridge-cli#388]
Everybody wants to manage cluster topology using Cartridge CLI. 

It's time to describe RFC for this great feature.

**Comment by dokshina** [comment]
## RFC

I suggest to implement `cartridge topology` command. 
This command should be used for local development (for production we have Ansible role).

I see these cases we need to support:

* Setting up topology described in a file (for example, `topology.yml`)
* Bump current topology into file
* Modify topology via `cartridge topology do-smth` commands. (Do we really need this? Maybe loading/saving topology to a file + Cartridge Web UI is enough?)
* Bootstrap vshard
* Set up failover (on the one hand by no, but on the other - it's quite complex operation that has many parameters, and I'm not sure that failover is always needed for local development)

### `topology.yml`

I would like this file not to be like Cartridge `topology.yml` (user shouldn't now anything about this file).
Template application contains `topology.yml` for instances described in `instances.yml`.

```yaml
# topology.yml
replicasets:
  router:
    instances:
    - router
    roles:
    - vshard-router
    - app.roles.custom
    - failover-coordinator
    - metrics
  s-1:
    instances:
    - s1-master
    - s1-replica
    roles:
    - vshard-storage
    - metrics
    weight: 1
    all_rw: false
    vshard_group: default
  s-2:
    instances:
    - s2-master
    - s2-replica
    roles:
    - vshard-storage
    - metrics
    weight: 1
    all_rw: false
    vshard_group: default
```

Since we have `instances.yml` that contains advertise URI for each instance, we can easily use instances aliases.

### Bootstrapping the cluster

So, now user can do

```bash
cartridge start -d
cartridge topology setup --bootstrap-vshard
```

### `topology` commands

#### Setting up topology

`cartridge topology setup` - sets up topology from file

* `--file` - file to load topology from (defaults to `topology.yml`)
*  `--run-dir` - directory where PID and socket files are stored (defaults to `tmp/run`)
* `--bootstrap-vshard` - additionally bootstrap vshard

If cluster is already bootstrapped, then current topology is changed (like Ansible role do).

#### Show current topology

Since we are working with the topology from console, we heed to look at it's current state sometimes.

`cartridge topology list` - shows current topology

*  `--run-dir` - directory where PID and socket files are stored (defaults to `tmp/run`)

Example:

```bash
cartridge topology list
The current topology is:
replicasets:
  router:
    instances:
    - router
    roles:
    - vshard-router
    - app.roles.custom
    - failover-coordinator
    - metrics
```



#### Save current topology to file

`cartridge topology save` - sets up topology from file

* `--file` - file to save topology  (defaults to `topology.yml`)
*  `--run-dir` - directory where PID and socket files are stored (defaults to `tmp/run`)

#### Bootstrap vshard

`cartridge topology bootstrap-vshard` - bootstrap vshard

*  `--run-dir` - directory where PID and socket files are stored (defaults to `tmp/run`)

#### Join/remove instance(s) to/from replicaset

`cartridge topology instances-join INSTANCE_NAME...`

`cartridge topology instances-remove INSTANCE_NAME...`

* `--replicaset` - replicaset alias
*  `--run-dir` - directory where PID and socket files are stored (defaults to `tmp/run`)

Example:

```bash
cartridge topology join --replicaset s-3 s3-master s3-replica
```

#### Add/remove role

`cartridge topology roles-add ROLE...`

`cartridge topology roles-remove ROLE...`

* `--replicaset` - replicaset alias
*  `--run-dir` - directory where PID and socket files are stored (defaults to `tmp/run`)

Example:

```bash
cartridge topology roles-add  --replicaset s-3 some-role other-role
```

#### Set failover priority

`cartridge topology failover-priority-set INSTANCE_NAME...`

* `--replicaset` - replicaset alias
*  `--run-dir` - directory where PID and socket files are stored (defaults to `tmp/run`)

Example:

```bash
cartridge topology failover-priority-set --replicaset s-3 s3-new-master s3-master
```

#### Set weight

`cartridge topology weight-set WEIGHT`

* `--replicaset` - replicaset alias
*  `--run-dir` - directory where PID and socket files are stored (defaults to `tmp/run`)

Example:

```bash
cartridge topology weight-set --replicaset s-3 2

#### Set vshard group??

`cartridge topology vshard-group-set VSHARD_GROUP`

* `--replicaset` - replicaset alias
*  `--run-dir` - directory where PID and socket files are stored (defaults to `tmp/run`)

Example:

```bash
cartridge topology vshard-group-set --replicaset s-3 hot
```

**Comment by dokshina** [comment]
Closed via #401
